From 097c528660a7a27c889ac943086d8284564a971b Mon Sep 17 00:00:00 2001
From: Dongwon Kim <dongwon.kim@intel.com>
Date: Thu, 18 Apr 2024 16:37:15 +0000
Subject: [PATCH 33/40] ui/gtk: gd_ui_show and hide only when ui_info exists

There are corner cases where we call gd_ui_show and hide
while the console doesn't have ui_info set. gd_ui_show and
hide should just return in such cases to prevent assert failure
(ui_info != NULL)

Signed-off-by: Dongwon Kim <dongwon.kim@intel.com>
---
 ui/gtk.c | 27 ++++++++++++++++++---------
 1 file changed, 18 insertions(+), 9 deletions(-)

diff --git a/ui/gtk.c b/ui/gtk.c
index b3bd7f3fcd37..b27d0eb4a096 100644
--- a/ui/gtk.c
+++ b/ui/gtk.c
@@ -867,6 +867,12 @@ static void gd_ui_hide(VirtualConsole *vc)
     QemuUIInfo info;
 
     vc->gfx.visible = false;
+    if (!dpy_ui_info_supported(vc->gfx.dcl.con)) {
+        /* forcefully cancel rendering sequence */
+        gd_hw_gl_flushed(vc);
+        return;
+    }
+
     info = *dpy_get_ui_info(vc->gfx.dcl.con);
     info.width = 0;
     info.height = 0;
@@ -881,6 +887,11 @@ static void gd_ui_show(VirtualConsole *vc)
     GtkDisplayState *s = vc->s;
     GdkWindow *window = gtk_widget_get_window(vc->gfx.drawing_area);
 
+    vc->gfx.visible = true;
+    if (!dpy_ui_info_supported(vc->gfx.dcl.con)) {
+        return;
+    }
+
     info = *dpy_get_ui_info(vc->gfx.dcl.con);
     info.width = gdk_window_get_width(window);
     info.height = gdk_window_get_height(window);
@@ -893,8 +904,6 @@ static void gd_ui_show(VirtualConsole *vc)
         gd_ungrab_keyboard(s);
         gd_ungrab_pointer(s);
     }
-
-    vc->gfx.visible = true;
 }
 
 #if defined(CONFIG_OPENGL)
@@ -2306,14 +2315,14 @@ static gboolean gd_configure(GtkWidget *widget,
 {
     VirtualConsole *vc = opaque;
 
-    if (vc->gfx.visible) {
-	QemuUIInfo info;
-	GdkWindow *window = gtk_widget_get_window(vc->gfx.drawing_area);
+    if (vc->gfx.visible && dpy_ui_info_supported(vc->gfx.dcl.con)) {
+        QemuUIInfo info;
+        GdkWindow *window = gtk_widget_get_window(vc->gfx.drawing_area);
 
-	info = *dpy_get_ui_info(vc->gfx.dcl.con);
-	info.width = gdk_window_get_width(window);
-	info.height = gdk_window_get_height(window);
-	dpy_set_ui_info(vc->gfx.dcl.con, &info, true);
+        info = *dpy_get_ui_info(vc->gfx.dcl.con);
+        info.width = gdk_window_get_width(window);
+        info.height = gdk_window_get_height(window);
+        dpy_set_ui_info(vc->gfx.dcl.con, &info, true);
     }
 
     return FALSE;
-- 
2.34.1

